Added ctranslate2 translation example script

uahmed93 opened the pull request:
Added a ctransalte2 example which works on string tokens instead of integer tokens.

**To run the example :**
```
python3 example/custom_ct2_model.py --ct2-model-dir <your-modelp-dir> inp.parquet out.parquet
```

uahmed93 commented:
This script is breaking because it is unable to serialize ct2 model while loading on worker. I have run this after initializing NDC gpu dask cluster on slurm but this is not working here.

VibhuJawa commented:
@uahmed93 , Can you post the error you saw here please.

uahmed93 commented:
Here:
```
Deployed LocalCUDACluster(22f3d4fd, 'tcp://127.0.0.1:37693', workers=1, threads=1, memory=1.79 TiB)...
2024-09-11 10:56:33,005 - distributed.protocol.pickle - ERROR - Failed to serialize <ToPickle: HighLevelGraph with 4 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x15522cf81900>
 0. read-parquet-2f062a30cf4676cd0b9a4ab5cf06fe85
 1. repartition-2-aa356a5cb4dedecfbdd625f941566718
 2. to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1
 3. store-to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1
>.
Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 63, in dumps
    result = pickle.dumps(x, **dump_kwargs)
AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 68, in dumps
    pickler.dump(x)
AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 81, in dumps
    result = cloudpickle.dumps(x, **dump_kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
TypeError: cannot pickle 'ctranslate2._ext.Translator' object
Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 63, in dumps
    result = pickle.dumps(x, **dump_kwargs)
AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 68, in dumps
    pickler.dump(x)
AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 353, in serialize
    header, frames = dumps(x, context=context) if wants_context else dumps(x)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 76, in pickle_dumps
    frames[0] = pickle.dumps(
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 81, in dumps
    result = cloudpickle.dumps(x, **dump_kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
TypeError: cannot pickle 'ctranslate2._ext.Translator' object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/ctransl/ctransl_cf.py", line 145, in <module>
    main()
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/ctransl/ctransl_cf.py", line 141, in main
    outputs.to_parquet(args.output_parquet_path)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/nvtx/nvtx.py", line 116, in inner
    result = func(*args, **kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask_cudf/core.py", line 264, in to_parquet
    return to_parquet(self, path, *args, **kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/dataframe/io/parquet/core.py", line 1047, in to_parquet
    out = out.compute(**compute_kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/base.py", line 379, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/base.py", line 665, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 379, in serialize
    raise TypeError(msg, str_x) from exc
TypeError: ('Could not serialize object of type HighLevelGraph', '<ToPickle: HighLevelGraph with 4 layers.\n<dask.highlevelgraph.HighLevelGraph object at 0x15522cf81900>\n 0. read-parquet-2f062a30cf4676cd0b9a4ab5cf06fe85\n 1. repartition-2-aa356a5cb4dedecfbdd625f941566718\n 2. to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1\n 3. store-to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1\n>')
```

VibhuJawa reviewed (changes requested):
Thanks, @uahmed93, for implementing this feature. The overall approach looks solid, but I've left some suggestions to enhance the code's readability and design.

The foundation is strong, but with a few refinements, it can be polished for better performance and maintainability.

VibhuJawa commented:
Also , CC: @sarahyurick for suggestions .

sarahyurick commented on crossfit/backend/torch/op/base.py:
```diff
@@ -81,27 +81,41 @@ def call(self, data, partition_info=None):
+                    all_outputs_ls.append(o)
+            else:
+                all_outputs_ls.append(output)
+        # print(f"all_outputs : {all_outputs_ls}")
```
```suggestion

```

uahmed93 replied:
Removed.

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,144 @@
+        self.string_tok_inf = config.string_tok_inf
+
+    def clean_extra_tokens(self, token_2d):
+        results=[]
```
```suggestion
        results = []
```
Can you run the `black` tool so that formatting mistakes like this will be fixed?

sarahyurick commented on crossfit/backend/torch/op/base.py:
```diff
@@ -81,27 +81,41 @@ def call(self, data, partition_info=None):
 
-            all_outputs_ls.append(output)
-
+            if self.model.string_tok_inf:
```
Why don't you just check if it's a `CT2CustomModel` (or check if it's a `TranslationConfig`, whichever makes more logical sense to you) instead of creating the `string_tok_inf` field?

uahmed93 replied:
Modified.

sarahyurick commented on crossfit/backend/torch/op/base.py:
```diff
@@ -81,27 +81,41 @@ def call(self, data, partition_info=None):
-        elif len(outputs.shape) == 3:
-            out[self.pred_output_col] = create_nested_list_series_from_3d_ar(outputs, _index)
+
+        if self.model.string_tok_inf:
```
Why don't you just check if it's a `CT2CustomModel` (or check if it's a `TranslationConfig`, whichever makes more logical sense to you) instead of creating the `string_tok_inf` field?

uahmed93 replied:
Modified

sarahyurick commented on crossfit/backend/torch/op/base.py:
```diff
@@ -81,27 +81,41 @@ def call(self, data, partition_info=None):
 
     def meta(self):
-        return {self.pred_output_col: "float32"}
+        if self.model.string_tok_inf:
```
Why don't you just check if it's a `CT2CustomModel` (or check if it's a `TranslationConfig`, whichever makes more logical sense to you) instead of creating the `string_tok_inf` field?

uahmed93 replied:
Now handling it with model_output_type attribute in base Model class.

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,144 @@
+class TranslationConfig:
+    pretrained_model_name_or_path: str
+    ct2_model_path: str
+    string_tok_inf: bool = True
```
When will `string_tok_inf` be False?

uahmed93 replied:
Removed this parameter.

uahmed93 commented on crossfit/backend/torch/op/base.py:
```diff
@@ -81,27 +81,41 @@ def call(self, data, partition_info=None):
+        if self.model.string_tok_inf:
+            out[self.pred_output_col] = cudf.Series(data=all_outputs_ls, index=_index)
+            del all_outputs_ls
+            del loader
```
Added an attribute to Model class : model_output_type, now handling output types with this.

uahmed93 commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,144 @@
+class TranslationConfig:
+    pretrained_model_name_or_path: str
+    ct2_model_path: str
+    string_tok_inf: bool = True
```
removed.

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,141 @@
+            trust_remote_code=True,
+        )
+        return config
+
```
Should have 2 empty lines between classes and functions.
```suggestion


```

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,141 @@
+
+    args = parser.parse_args()
+    return args
+
```
Should have 2 empty lines between functions.
```suggestion


```

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,141 @@
+    with cf.Distributed(rmm_pool_size=args.pool_size, n_workers=args.num_workers):
+        model = ModelForSeq2SeqModel(Config)
+        pipe = op.Sequential(
+            op.Tokenizer(model, cols=[args.input_column], tokenizer_type="sentencepiece", max_length=255),
```
```suggestion
            op.Tokenizer(
                model, cols=[args.input_column], tokenizer_type="sentencepiece", max_length=255
            ),
```

sarahyurick commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,141 @@
+
+
+if __name__ == "__main__":
+    main()
```
```suggestion
    main()

```

sarahyurick reviewed (changes requested):
Thanks @uahmed93 for the updates! I ran `black` and there are several lines which need to be reformatted, so I added them here.

sarahyurick commented:
Also @uahmed93, I was wondering if I could test this PR myself by downloading the model files at https://huggingface.co/ai4bharat/indictrans2-en-indic-1B? Or should I be using something else?

uahmed93 commented:
Thanks @VibhuJawa  and @sarahyurick for detailed analysis. I have now added get_model_output function in Model class to handle postprocessing of outputs as suggested by @VibhuJawa .
I have run black for formatting now.

uahmed93 commented:
@sarahyurick  that is hf model. CT2 model can be downloaded from [here](https://github.com/AI4Bharat/IndicTrans2/tree/main#download-models-and-other-artifacts) Base En-indic model.
After downloading you need to unzip and there will be ct2_fp16 model. That model path need to be provided.

sarahyurick reviewed (approved):
Tested and it worked for me. Thanks @uahmed93 !

VibhuJawa commented on crossfit/backend/torch/model.py:
```diff
@@ -41,3 +56,41 @@ def estimate_memory(self, max_num_tokens: int, batch_size: int) -> int:
+        out = cudf.DataFrame(index=index)
+        _index = (
+            loader.sort_column(index.values)
+            if loader.__class__.__name__ == "SortedSeqLoader"
+            else index
```
This is not recommended. Please use isinstace.

uahmed93 replied:
I have a doubt here, this sorting of index required only for SortedSeqLoader and not for InMemoryLoader, right ?
If this is the case, then isinstance will return True even for InMemoryLoader loader object.

VibhuJawa replied:
```python3
class Base:
    pass

class Derived(Base):
    pass

obj1 = Base()
obj2 = Derived()

# Check for exact instance of Base, not matching subclasses
if type(obj1) is Base:
    print("obj1 is exactly of type Base")

if type(obj2) is Base:
    print("obj2 is exactly of type Base")
else:
    print("obj2 is not exactly of type Base")

# Check using isinstance, which includes subclasses
if isinstance(obj1, Base):
    print("obj1 is an instance of Base or its subclass")

if isinstance(obj2, Base):
    print("obj2 is an instance of Base or its subclass")
 ```
 
 ```python3
 obj1 is exactly of type Base
obj2 is not exactly of type Base
obj1 is an instance of Base or its subclass
obj2 is an instance of Base or its subclass
 ```
 
I think if you want to match subclasses , you can use type instead.

uahmed93 replied:
Done

VibhuJawa commented on crossfit/backend/torch/model.py:
```diff
@@ -11,10 +11,25 @@
+    def __init__(self, path_or_name: str, max_mem_gb: int = 16, model_output_type: str = "numeric"):
         self.path_or_name = path_or_name
         self.max_mem_gb = max_mem_gb
+        if model_output_type == "numeric" or model_output_type == "string":
```
```suggestion
        if model_output_type in ["numeric", "string"]:
```

uahmed93 replied:
Done.

VibhuJawa reviewed (changes requested):
Please add  pytests, mostly other things look good to me.

VibhuJawa commented on crossfit/backend/torch/model.py:
```diff
@@ -41,3 +56,41 @@ def estimate_memory(self, max_num_tokens: int, batch_size: int) -> int:
+        )
+
+        if self.model_output_type == "string":
+            all_outputs = []
+            for output in all_outputs_ls:
+                for o in output:
+                    all_outputs.append(o)
```
Cleanup this code
```suggestion
all_outputs = [o for output in all_outputs_ls for o in output]
```

uahmed93 replied:
Handled.

VibhuJawa commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,155 @@
+import argparse
```
Add copy right from Nvidia in the starting

```
# Copyright 2024 NVIDIA CORPORATION
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

VibhuJawa commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,155 @@
+        model = ModelForSeq2SeqModel(Config)
+        pipe = op.Sequential(
+            op.Tokenizer(
+                model, cols=[args.input_column], tokenizer_type="sentencepiece", max_length=255
```
```suggestion
                model, cols=[args.input_column], tokenizer_type="default", max_length=255
```

uahmed93 replied:
changed.

uahmed93 commented:
Hi, @VibhuJawa  @sarahyurick 
Sorry was doing some other priority task due to which I could not handle the reviews immediately.
I have added test case here and corrected some suggestions. can please review them?.

VibhuJawa commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,170 @@
+            tokens_1d[i : i + token_ids_2d.size(1)]
+            for i in range(0, len(tokens_1d), token_ids_2d.size(1))
+        ]
+        tokenss = self.clean_extra_tokens(tokens_2d)
```
Replace `tokenss` with `tokens` everywhere

uahmed93 replied:
Replaced.

VibhuJawa commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,170 @@
+        return model
+
+    def load_config(self):
+        return AutoConfig.from_pretrained(
+            pretrained_model_name_or_path=self.trans_config.pretrained_model_name_or_path,
+            trust_remote_code=True,
+        )
```
Use the load_cfg function, dont repeat code.

uahmed93 replied:
Using load_cfg() now

VibhuJawa commented on examples/custom_ct2_model.py:
```diff
@@ -0,0 +1,162 @@
+        self.model = ctranslate2.Translator(model_path=config.ct2_model_path, device=device)
+
+    def clean_extra_tokens(self, token_2d):
+        results = []
+        for token_1d in token_2d:
+            result = []
+            for t in token_1d:
+                if (
+                    t == self.tokenizer.pad_token
+                    or t == self.tokenizer.bos_token
+                    or t == self.tokenizer.eos_token
+                    or t == self.tokenizer.unk_token
+                ):
+                    pass
+                else:
+                    result.append(t)
+            results.append(result)
```
```suggestion
        results = [
    [t for t in token_1d if t not in {self.tokenizer.pad_token, self.tokenizer.bos_token, self.tokenizer.eos_token, self.tokenizer.unk_token}]
    for token_1d in token_2d
]
```

VibhuJawa commented on tests/op/test_model_function.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@pytest.mark.parametrize("trust_remote_code", ["y"])
+def test_model_output_int(trust_remote_code, model_name="ai4bharat/indictrans2-en-indic-1B"):
```
As we are not really using the `ai4bharat/indictrans2-en-indic-1B` here , can we just switch this to a diffrent model like `microsoft/deberta-v3-base`.

uahmed93 replied:
ok let me test it

VibhuJawa commented on tests/op/test_model_function.py:
```diff
@@ -0,0 +1,67 @@
+        pred_output_col = "translation"
+        out = model.get_model_output(all_outputs_ls, index, loader, pred_output_col)
+        assert isinstance(out, cudf.DataFrame)
+        assert isinstance(out["translation"][0][0], int)
```
I will also test that the out["translation"] has the expected values.

uahmed93 replied:
added a check of individual values

VibhuJawa commented on tests/op/test_model_function.py:
```diff
@@ -0,0 +1,67 @@
+        pred_output_col = "translation"
+        out = model.get_model_output(all_outputs_ls, index, loader, pred_output_col)
+        assert isinstance(out, cudf.DataFrame)
+        assert isinstance(out["translation"][0][0], str)
```
I will also test that the out["translation"] has the expected values.

VibhuJawa reviewed (changes requested):
Last final bits  of review which i missed in the initial review, we should be good to merge after these are fixed.

uahmed93 commented:
> Here:
> 
> ```
> Deployed LocalCUDACluster(22f3d4fd, 'tcp://127.0.0.1:37693', workers=1, threads=1, memory=1.79 TiB)...
> 2024-09-11 10:56:33,005 - distributed.protocol.pickle - ERROR - Failed to serialize <ToPickle: HighLevelGraph with 4 layers.
> <dask.highlevelgraph.HighLevelGraph object at 0x15522cf81900>
>  0. read-parquet-2f062a30cf4676cd0b9a4ab5cf06fe85
>  1. repartition-2-aa356a5cb4dedecfbdd625f941566718
>  2. to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1
>  3. store-to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1
> >.
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 63, in dumps
>     result = pickle.dumps(x, **dump_kwargs)
> AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 68, in dumps
>     pickler.dump(x)
> AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 81, in dumps
>     result = cloudpickle.dumps(x, **dump_kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1479, in dumps
>     cp.dump(obj)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
>     return super().dump(obj)
> TypeError: cannot pickle 'ctranslate2._ext.Translator' object
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 63, in dumps
>     result = pickle.dumps(x, **dump_kwargs)
> AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 68, in dumps
>     pickler.dump(x)
> AttributeError: Can't pickle local object 'to_parquet.<locals>.<lambda>'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 353, in serialize
>     header, frames = dumps(x, context=context) if wants_context else dumps(x)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 76, in pickle_dumps
>     frames[0] = pickle.dumps(
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 81, in dumps
>     result = cloudpickle.dumps(x, **dump_kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1479, in dumps
>     cp.dump(obj)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
>     return super().dump(obj)
> TypeError: cannot pickle 'ctranslate2._ext.Translator' object
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/ctransl/ctransl_cf.py", line 145, in <module>
>     main()
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/ctransl/ctransl_cf.py", line 141, in main
>     outputs.to_parquet(args.output_parquet_path)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/nvtx/nvtx.py", line 116, in inner
>     result = func(*args, **kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask_cudf/core.py", line 264, in to_parquet
>     return to_parquet(self, path, *args, **kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/dataframe/io/parquet/core.py", line 1047, in to_parquet
>     out = out.compute(**compute_kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/base.py", line 379, in compute
>     (result,) = compute(self, traverse=False, **kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/dask/base.py", line 665, in compute
>     results = schedule(dsk, keys, **kwargs)
>   File "/lustre/fsw/portfolios/llmservice/users/uahmed/test_env/lib/python3.10/site-packages/distributed/protocol/serialize.py", line 379, in serialize
>     raise TypeError(msg, str_x) from exc
> TypeError: ('Could not serialize object of type HighLevelGraph', '<ToPickle: HighLevelGraph with 4 layers.\n<dask.highlevelgraph.HighLevelGraph object at 0x15522cf81900>\n 0. read-parquet-2f062a30cf4676cd0b9a4ab5cf06fe85\n 1. repartition-2-aa356a5cb4dedecfbdd625f941566718\n 2. to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1\n 3. store-to-parquet-03d4d012cafd9ed3bdb4ee7374761cf1\n>')
> ```

Resolved it.

VibhuJawa commented on crossfit/backend/torch/model.py:
```diff
@@ -41,3 +58,36 @@ def estimate_memory(self, max_num_tokens: int, batch_size: int) -> int:
         raise NotImplementedError()
+
+    def get_model_output(self, all_outputs_ls, index, loader, pred_output_col) -> cudf.DataFrame:
+        from crossfit.backend.torch.loader import SortedSeqLoader
```
Sorry missed this earlier, we should move the import up.

uahmed93 replied:
That was creating cyclic import problem.

VibhuJawa replied:
Just add a comment here.

uahmed93 replied:
Sure.

VibhuJawa commented on tests/op/test_model_function.py:
```diff
@@ -0,0 +1,77 @@
+        out = model.get_model_output(all_outputs_ls, index, loader, pred_output_col)
+        assert isinstance(out, cudf.DataFrame)
+        assert isinstance(out["translation"][0][0], int)
+        assert (
+            out["translation"][0] == data[2]
+            and out["translation"][1] == data[1]
+            and out["translation"][2] == data[0]
```
Why is this reverse ?

uahmed93 replied:
```
self.sorted_indices = seq_length.argsort(descending=True)
```
[this](https://github.com/uahmed93/crossfit/blob/22844c7d19901b5e3a582e9397a25060237e5f84/crossfit/backend/torch/loader.py#L123) line was creaing problem, even if lengths were same which were [3,3,3], indexes were sorted in descending manner, like [2,1,0].
Anyway now changed the input_ids from function.

VibhuJawa commented on tests/op/test_model_function.py:
```diff
@@ -0,0 +1,77 @@
+        out = model.get_model_output(all_outputs_ls, index, loader, pred_output_col)
+        assert isinstance(out, cudf.DataFrame)
+        assert isinstance(out["translation"][0][0], str)
+        assert (
+            out["translation"][0] == data[2][0]
+            and out["translation"][1] == data[1][0]
+            and out["translation"][2] == data[0][0]
```
Same question here ?

uahmed93 replied:
Changed

VibhuJawa reviewed (changes requested):
Last nits around testing

VibhuJawa commented:
/okay to test

VibhuJawa commented:
@uahmed93 , Please fix linting issue.

VibhuJawa commented:
/okay to test
